### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by paullferguson** [mechio/takana#33]
Big ask I know, but might as well gauge interest.
I still shed a tear for the days of OSX and CSSEdit.

**Comment by Talv** [comment]
I got it to work. Takana itself has no problems runing on windows. But it uses two tools from linux based systems - **find** and **rsync**. So get the equivalents ports of them for windows, and make sure they are available from windows PATH (you will also need to override default windows **find** cmd). Rsync is used only at start of the takana to copy all the scss and css files from your project to takana temp directory. Find is used to list all the scss and css files in project directory to let takana hook them which will allow it to receive events about modifications.

Anyway, it's awesome tool, it was worth to get through those issues.

**Comment by barnaby** [comment]
Awesome that you got it working @Talv! Any chance you could link to the find and rsync ports that worked for you?

Did you not have to change any of the code at all? I always assumed that you'd have to change some of the paths.

**Comment by Talv** [comment]
I've used find from msysGit package https://msysgit.github.io/
With rsync it isn't as easy though. As there's no native port for windows. Atleast known to me. There are few that are based on cygwin. I choosed [DeltaCopy](http://www.aboutmyip.com/AboutMyXApp/DeltaCopy.jsp) cause it is shipped with cygwin runtime libraries. So you don't need to have cygwin installed on your system.
The only one modification to the code i had to do was related to rsync. The port i used wasn't setting up files permissions on default options set. Resolved it by putting ```--chmod=ugo=rwX```. Also it requires cygwin like path - ```/cygwin/c/Users/```. So i've ended up with hack like this:
```coffee
# lib/watcher/folder.coffee
cmd = "rsync --chmod=ugo=rwX -arq --delete --copy-links --exclude='node_modules/' --exclude='.git' --include='+ */' #{includes} --exclude='- *' ./ '/cygdrive/c/Users/KK/.takana/scratch/'"
```
Plus you need to install sublime text plugin manually (via package control). Cause there's no installation path for windows listed in takana obviously.
And that's it - nothing more was needed :)

**Comment by barnaby** [comment]
Thanks, this is really interesting. 

Are you running takana from cygwin or the windows command prompt?

My original thought was to re-implement `rsync` and `find` in native javascript, but would definitely anticipate a performance hit. I wonder if we could just package these up inside the takana repo instead.

Downloading Windows now, going to have a look and see how easy it might be.

**Comment by Talv** [comment]
I don't have cygwin installed at all. I run takana from BASH emulation that is in msysGit package. But it works from windows command prompt aswell.
I think it is definitely possible to just package up rsync and find binaries into takana. Just provide absolute path to those in the code, assuming that they won't be added into windows PATH. Also rsync will need special treatment like i mentioned before, translating pathes from ```c:\Users\``` to ```/cygdrive/c/Users```. While leaving standard pathes for rest of the code.

**Comment by dawjan** [comment]
Its working on windows:
Just create symlink:

http://geeknizer.com/create-symlink-in-windows/

C:\Users\[USER_NAME]\undefined\Library\Application support\ [ Sublime Text 3 => to sublime dir ]

But only on chroma that strange
![asd](https://cloud.githubusercontent.com/assets/908583/5893037/d5f2490e-a4cd-11e4-856f-dbba6aebf767.png)

**Comment by MattyBalaam** [comment]
I have got quite far, but I assume if I only get 'buffer updated', and do not get 'rendered' in the cli that I need to make sure I manually install windows versions of rsync and find? If anyone has some easy to follow steps for this I would appreciate it.

**Comment by Talv** [comment]
Rsync and find aren't really needed if you work with just a single SCSS file. Cause then you can provide content of the file via sublime plugin - 'buffer update' message. 
I assume that part works good for you.
To diagnose why your stylesheet isn't rendered, firstly run takana in verbose mode (```-v``` option). Refresh the browser, and check the log:
```
[INFO] BrowserManager - browser connected to project default
```
If you don't get the above it means that the javascript client of takana didin't run correctly - check the console for errors, make sure you have it included.
The next message you should see is:
```
[TRACE] BrowserManager - received event: 'stylesheet:resolve'
```
That's the request from the takana client to resolve stylesheet path and match it against the filesystem path.
This message should be followed by:
```
[INFO] Server - matched
```
If you don't get that one (i bet this is the case) it means that takana couldn't match the web server path (http://localhost/example/style.css) against file system path (c:\example\style.scss).
Note that windows uses backslash instead of forward slash. That's the issue most likely, cause takana uses regexps prepared to match "/" not "\".
Although, i remember it somehow worked for me without tweaking anything in this area. I'm no longer on windows so can't test it at the moment.
But what you can do for the begining is make a project where stylesheet is in the root directory (http://localhost/style.css). And see if that helps.

**Comment by MattyBalaam** [comment]
Excellent rundown. thank you very much.

I never get as far as the second BrowserManager message, I get a couldn't find a match error when I refresh the page. I have tried every combination of back and forward slashes I could think of when starting the project:

c:/MAMP/htdocs/test/
c:/MAMP/htdocs/test
c:\MAMP\htdocs\test\
c:\MAMP\htdocs\test

Here is an example log:

[2015-02-25 10:35:16.603] [WARN] [default] - if you have just upgraded from vers
ion 0.0.10 or below, please update the javascript snippet (above)
[2015-02-25 10:35:16.608] [DEBUG] Server - Staring...
[2015-02-25 10:35:16.609] [DEBUG] Server - starting rsync
[2015-02-25 10:35:17.265] [INFO] EditorManager - editor server listening on 48627
[2015-02-25 10:35:17.267] [INFO] Server - webserver listening on 48626
[2015-02-25 10:35:17.283] [DEBUG] Server - rsync finished
[2015-02-25 10:35:17.306] [DEBUG] Server - started
[2015-02-25 10:35:17.308] [ERROR] [default] - couldn't find a Sublime Text package directory. Install an editor plugin manually.
[2015-02-25 10:35:17.310] [ERROR] [default] - search paths were: [ 'c:\\MAMP\\htdocs\\test\\undefined\\Library\\Application Support\\Sublime Text 3\\Packages/',
  'c:\\MAMP\\htdocs\\test\\undefined\\Library\\Application Support\\Sublime Text
 2\\Packages/',  'c:\\MAMP\\htdocs\\test\\undefined\\.config\\sublime-text-3\\Packages/' ]
[2015-02-25 10:35:17.314] [INFO] [default] - running on project c:\MAMP\htdocs\test
[2015-02-25 10:35:22.790] [INFO] BrowserManager - browser connected to project default
[2015-02-25 10:35:22.793] [TRACE] BrowserManager - received event: 'stylesheet:resolve', data: {href: 'http://localhost/test/css/styles.css' }
[2015-02-25 10:35:22.796] [WARN] Server - couldn't find a match for http://localhost/test/css/styles.css
[2015-02-25 10:35:22.807] [TRACE] BrowserManager - received event: 'stylesheet:listen', data: {}
[2015-02-25 10:35:22.809] [DEBUG] Server - processing stylesheet:listen undefined
[2015-02-25 10:35:32.473] [INFO] EditorManager - editor connected
[2015-02-25 10:35:35.094] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:38.731] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:38.865] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:39.092] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:39.255] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:39.770] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:40.008] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss
[2015-02-25 10:35:40.091] [INFO] EditorManager - buffer updated for C:\MAMP\htdocs\test\scss\styles.scss

**Comment by Talv** [comment]
Basing on the log it appears that you refresh the browser as first and then you are updating the buffer for the file. It should be in opposite order :) Takana isn't aware about the file until you update it's buffer by sublime plugin. 
As for the slashes, providing path with "/" for takana as a project directory won't change anything unfortunately. It will operate on "\" anyway, as it is native for windows and i doubt there's anything you can do about it.
I think the issue lies within file_matcher.coffee , and his regexps designed to match "/". Although it uses a couple of methods to match the files, not every of them based on regexps.
https://github.com/mechio/takana/blob/master/lib/support/file_matcher.coffee#L63
That's why it may work in some cases and in others just not at all..
So, if updating the buffer of the file before refreshing the browser won't solve the problem, then you may take a shot at tweaking the code or just wait for official windows version from barnaby :)

**Comment by MattyBalaam** [comment]
Thank you very much again. I can't still get this to work, but I'm sure you will help someone else figure it out. Hopefully an official windows version is on the way :)

**Comment by hydrosquall** [comment]
@dawjan Thanks for the Symlink suggestion. Although I couldn't get it to work, it inspired me to go to the existing folder
    C:\Users\MYUSERNAME\.config
I then added 2 folders and pasted Takana.sublime-package into it. (I had gotten the sublime package from package manager).
    C:\Users\MYUSERNAME\.config\sublime-text-3\package

Unfortunately, while Takana succeeds at launching the editor server + webserver, and the browser connects to the project default, it's been unable to find a match for the 2 CSS files listed in the head.

In other words, I'm only getting to the first BrowserManager message, and failing the subsequent "match css" message above. I tried putting the styles in the root directoryy of the project, and it didn't help.

@Talv , I installed Git for Windows which [this thread](C:\Users\Cameron\Documents\GitHub\sass_step) says has merged with mysysgit. So Find should be taken care of. 

 Would installing rsync help, or is the crux of the issue the regex matching of the slashes?

**Comment by sanjeev18** [comment]
Hello there, I am able to update buffer and i think i know the reason it's not working on my pc.

let me breakdown what i did in steps.

1. Installed Node.js
2. Installed takana using npm install -g takana
3. installed msysgit
4. installed DeltaCopy

5. As there is only installation location of sublme plugin for linux and mac. I have edited index.js and manually added installation location for windows.
![image](https://cloud.githubusercontent.com/assets/2367137/12383146/95b9639c-bdca-11e5-9a93-ac6b1daacd7f.png)

6. Ran takana in verbose mode and got buffered update
![image](https://cloud.githubusercontent.com/assets/2367137/12383173/fbfd970e-bdca-11e5-87d4-6c19aeef2970.png)

Now i couldn't get it to rendered. I think problem lies on webport as i can't view takana.js
![image](https://cloud.githubusercontent.com/assets/2367137/12383218/6ba08bfc-bdcb-11e5-9bda-08dd1b405b36.png)

Maybe the snippet is not working because takana.js is not available, Anyone has a solution to that ?

Thank you

**Comment by opticoverload** [comment]
Having the same issue as @sanjeev18. Tried shutting down locally installed Apache in case of a potential conflict, but to no avail.

**Comment by sanjeev18** [comment]
@opticoverload .. I have found something else. Try livestyle it has all the features we need.

Thanks
